// yiffy/build/src/util/get.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Constants_1 = require("./Constants");
const https_1 = __importDefault(require("https"));
const url_1 = require("url");
async function get(url, ua, auth, host) {
    return new Promise((a, b) => {
        const u = new url_1.URL(url);
        https_1.default
            .get({
            host: u.host,
            path: `${u.pathname}${u.search}`,
            port: url.indexOf("https") !== -1 ? 443 : 80,
            method: "GET",
            headers: {
                "User-Agent": ua,
                "Host": host || Constants_1.API_HOST,
                ...(auth ? ({
                    Authorization: auth
                }) : ({}))
            }
        }, (res) => {
            const d = [];
            res
                .on("data", d.push.bind(d))
                .on("error", (err) => b(err))
                .on("end", () => a({
                headers: res.headers,
                body: Buffer.concat(d),
                statusCode: res.statusCode,
                statusMessage: res.statusMessage
            }));
        })
            .end();
    });
}
exports.default = get;
